hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/HoodieDatasetBulkInsertHelper.scala

```diff
@@ -150,14 +149,8 @@ object HoodieDatasetBulkInsertHelper
                  table: HoodieTable[_, _, _, _],
                  writeConfig: HoodieWriteConfig,
                  arePartitionRecordsSorted: Boolean,
-                 shouldPreserveHoodieMetadata: Boolean,
-                 operation: WriteOperationType): HoodieData[WriteStatus] = {
-    val schema = operation match {
-      case WriteOperationType.CLUSTER =>
-        alignNotNullFields(dataset.schema, new Schema.Parser().parse(writeConfig.getSchema))
-      case _ =>
-        dataset.schema
-    }
+                 shouldPreserveHoodieMetadata: Boolean): HoodieData[WriteStatus] = {
+    val schema = AvroConversionUtils.alignFields(dataset.schema, new Schema.Parser().parse(writeConfig.getSchema))
```

Comment:
Is this a enforcement for nexted data types?

Reply:
This data class correction is necessary. The existing "bulk insert as row" operation fetches the DataFrame's schema(may inconsistencies with table schema). This inconsistencies in the nullable properties of identical columns lead to file corruption/unreadability after binary copying. Therefore, we've implemented a nullable type correction for the bulk insert as row operation to resolve this issue.